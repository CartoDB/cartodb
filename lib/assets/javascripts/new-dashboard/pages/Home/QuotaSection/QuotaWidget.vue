<template>
  <div class="quota-widget" @mouseover="onMouseOver" @mouseleave="onMouseLeave">
    <div class="quota-main">
      <div class="quota-cell cell--title">
        <h4 class="text is-caption is-semibold is-txtGrey">{{name}}</h4>
      </div>
      <div class="quota-cell cell--large">
        <div class="progressbar">
            <div :class="`progressbar progressbar--${getStatusBar}`"  :style="{width: `${getUsedPercent}%`}">
            </div>
        </div>
      </div>
    </div>

    <div class="quota-data">
      <div class="quota-cell cell--medium">
        <span class="text is-small is-txtSoftGrey">{{roundOneDecimal(remainingQuota)}} {{unit}}</span>
      </div>
      <div class="quota-cell cell--medium cell--mobile">
        <span class="text is-small is-txtSoftGrey">{{roundOneDecimal(usedQuota)}} {{unit}}</span>
      </div>
      <div class="quota-cell cell--medium">
        <span class="text is-small is-txtSoftGrey">{{roundOneDecimal(availableQuota)}} {{unit}}</span>
      </div>
      <div class="quota-help cell--small">
        <a href=""><img svg-inline class="quota-image" :class="{'is-active': active}" src="../../../assets/icons/common/question-mark.svg"/></a>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'QuotaCard',
  props: {
    name: String,
    availableQuota: Number,
    usedQuota: Number,
    unit: String,
    billingPeriod: String
  },
  data: function () {
    return {
      active: false
    };
  },
  computed: {
    remainingQuota () {
      return this.availableQuota - this.usedQuota;
    },
    getUsedPercent () {
      if (this.availableQuota === 0) {
        return 100;
      }
      const width = (this.usedQuota / this.availableQuota) * 100;
      return width;
    },
    getStatusBar () {
      const usedPercent = this.getUsedPercent;
      if (usedPercent > 90) {
        return 'problem';
      } else if (usedPercent > 60) {
        return 'warning';
      } else {
        return 'good';
      }
    }
  },
  methods: {
    roundOneDecimal (number) {
      return Math.round(number * 10) / 10;
    },
    onMouseOver () {
      this.active = true;
    },
    onMouseLeave () {
      this.active = false;
    }
  }
};
</script>

<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.quota-widget {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 80px;
  cursor: pointer;

  &:not(:last-of-type) {
    border-bottom: 1px solid $softblue;
  }
}

.quota-main {
  display: flex;
  width: 100%;
}

.quota-data {
  display: flex;
}

.quota-cell {
  display: flex;
  align-items: center;
  padding: 0 10px;
  white-space: nowrap;

  &:first-of-type {
    padding-left: 0;
  }

  &:last-of-type {
    padding-right: 0;
  }
}

.cell--mobile {
  @media (max-width: $layout-mobile) {
    display: none;
  }
}

.quota-help {
  display: flex;
  justify-content: flex-end;
  padding-right: 16px;
}

.quota-image {
  display: none;

  &.is-active {
    display: flex;
  }
}

.cell--title {
  width: 110px;
  margin-left: 36px;
}

.cell--large {
  flex-grow: 1;
  flex-shrink: 1;
  min-width: 120px;
  margin-right: 40px;

  @media (max-width: $layout-mobile) {
    display: none;
  }
}

.cell--medium {
  width: 120px;
}

.cell--small {
  width: 58px;

  @media (max-width: $layout-tablet) {
    display: none;
  }
}

.progressbar {
  width: 100%;
  max-width: 240px;
  height: 8px;
  border-radius: 4px;
  background-color: $light-grey;

  &.progressbar--good {
    background-color: $good-state;
  }

  &.progressbar--warning {
    background-color: $warning-state;
  }

  .progressbar--problem {
    background-color: $error-state;
  }
}

</style>
