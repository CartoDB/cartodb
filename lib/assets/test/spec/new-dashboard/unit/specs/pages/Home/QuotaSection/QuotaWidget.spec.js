import { shallowMount } from '@vue/test-utils';
import QuotaWidget from 'new-dashboard/pages/Home/QuotaSection/QuotaWidget';

describe('QuotaWidget.vue', () => {
  it('should render correct contents', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'Widget test',
        quotaType: 'quota',
        availableQuota: 1234,
        usedQuota: 40,
        unit: 'GB',
        helpLink: 'https://test-help-link.com'
      }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  it('should render correctly compact widget', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'Another test',
        quotaType: 'credits',
        availableQuota: 1234,
        usedQuota: 12,
        mode: 'compact',
        helpLink: 'https://test-help-link.com'
      }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  it('should return 0 if the storage remaining quota is negative', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'Another test',
        quotaType: 'quota',
        availableQuota: 262144000,
        usedQuota: 289722368,
        unit: 'MB'
      }
    });

    let remainingQuota = quotaWidget.vm.remainingQuota;
    expect(remainingQuota).toBe(0);
  });

  it('should render warning bar if used ammount is over 60% and under 90%', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'My test',
        quotaType: 'test',
        availableQuota: 1000,
        usedQuota: 610,
        helpLink: 'https://test-help-link.com'
      }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  it('should render problem bar if used ammount is over 90%', () => {
    const quotaWidget = shallowMount(QuotaWidget, {
      propsData: {
        name: 'My test',
        quotaType: 'test',
        availableQuota: 1000,
        usedQuota: 932,
        helpLink: 'https://test-help-link.com'
      }
    });

    expect(quotaWidget).toMatchSnapshot();
  });

  describe('Computed', () => {
    it('should return correct percent', () => {
      const quotaWidget = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          quotaType: 'test',
          availableQuota: 350,
          usedQuota: 224
        }
      });

      const percent = quotaWidget.vm.getUsedPercent;

      expect(percent).toBe(64);
    });

    it('should not crash and return 100 when the available capacity is 0', () => {
      const quotaWidget = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          quotaType: 'test',
          availableQuota: 0,
          usedQuota: 0
        }
      });

      const percent = quotaWidget.vm.getUsedPercent;

      expect(percent).toBe(100);
    });

    it('should return correct status of the quota', () => {
      const quotaWidgetGood = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          quotaType: 'test',
          availableQuota: 100,
          usedQuota: 50
        }
      });
      const quotaWidgetWarning = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          quotaType: 'test',
          availableQuota: 100,
          usedQuota: 61.4
        }
      });
      const quotaWidgetProblem = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          quotaType: 'test',
          availableQuota: 100,
          usedQuota: 90.1
        }
      });

      const statusGood = quotaWidgetGood.vm.getStatusBar;
      const statusWarning = quotaWidgetWarning.vm.getStatusBar;
      const statusProblem = quotaWidgetProblem.vm.getStatusBar;

      expect(statusGood).toBe('good');
      expect(statusWarning).toBe('warning');
      expect(statusProblem).toBe('problem');
    });
  });

  describe('Methods', () => {
    it('should round decimals correctly correct percent', () => {
      const quotaWidget = shallowMount(QuotaWidget, {
        propsData: {
          name: 'My test',
          quotaType: 'test',
          availableQuota: 350,
          usedQuota: 224
        }
      });

      const value1 = quotaWidget.vm.roundOneDecimal(0);
      const value2 = quotaWidget.vm.roundOneDecimal(0.1234);
      const value3 = quotaWidget.vm.roundOneDecimal(120.2);
      const value4 = quotaWidget.vm.roundOneDecimal(129.0);
      const value5 = quotaWidget.vm.roundOneDecimal(45.8);
      const value6 = quotaWidget.vm.roundOneDecimal(45.26);

      expect(String(value1)).toBe('0');
      expect(String(value2)).toBe('0.1');
      expect(String(value3)).toBe('120.2');
      expect(String(value4)).toBe('129');
      expect(String(value5)).toBe('45.8');
      expect(String(value6)).toBe('45.3');
    });
  });
});
