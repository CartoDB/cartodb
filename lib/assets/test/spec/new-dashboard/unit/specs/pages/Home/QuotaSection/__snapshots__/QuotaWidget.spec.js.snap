// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotaWidget.vue should render correct contents 1`] = `
<div class="quota-widget" quotatype="quota">
  <div class="quota-main">
    <div class="quota-cell cell--title">
      <h4 class="text is-caption is-semibold is-txtGrey">Widget test</h4>
    </div>
    <div class="quota-cell cell--large">
      <div class="progressbar">
        <div class="progressbar progressbar--good" style="width: 3.2414910858995136%;"></div>
      </div>
    </div>
  </div>
  <div class="quota-data">
    <div class="quota-cell cell--medium"><span class="text is-small is-txtSoftGrey">
        1,194 GB
      </span></div>
    <div class="quota-cell cell--medium cell--mobile"><span class="text is-small is-txtSoftGrey">
        40 GB
      </span></div>
    <div class="quota-cell cell--medium"><span class="text is-small is-txtSoftGrey">
        1,234 GB
      </span></div>
    <div class="quota-help cell--small"><a href="https://test-help-link.com" target="_blank"><img svg-inline="" src="../../../assets/icons/common/question-mark.svg" class="quota-image"></a></div>
  </div>
</div>
`;

exports[`QuotaWidget.vue should render correctly compact widget 1`] = `
<div class="quota-widget" quotatype="credits" mode="compact">
  <div class="quota-main">
    <div class="quota-cell cell--title">
      <h4 class="text is-caption is-semibold is-txtGrey">Another test</h4>
    </div>
    <div class="quota-cell cell--large">
      <div class="progressbar">
        <div class="progressbar progressbar--good" style="width: 0.9724473257698543%;"></div>
      </div>
    </div>
  </div>
  <div class="quota-data">
    <div class="quota-cell cell--medium"><span class="text is-small is-txtSoftGrey">
        1,222 
      </span></div>
    <div class="quota-cell cell--medium cell--mobile"><span class="text is-small is-txtSoftGrey">
        12 
      </span></div>
    <div class="quota-cell cell--medium"><span class="text is-small is-txtSoftGrey">
        1,234 
      </span></div>
    <div class="quota-help cell--small"><a href="https://test-help-link.com" target="_blank"><img svg-inline="" src="../../../assets/icons/common/question-mark.svg" class="quota-image"></a></div>
  </div>
</div>
`;

exports[`QuotaWidget.vue should render problem bar if used ammount is over 90% 1`] = `
<div class="quota-widget" quotatype="test">
  <div class="quota-main">
    <div class="quota-cell cell--title">
      <h4 class="text is-caption is-semibold is-txtGrey">My test</h4>
    </div>
    <div class="quota-cell cell--large">
      <div class="progressbar">
        <div class="progressbar progressbar--problem" style="width: 93.2%;"></div>
      </div>
    </div>
  </div>
  <div class="quota-data">
    <div class="quota-cell cell--medium"><span class="text is-small is-txtSoftGrey">
        68 
      </span></div>
    <div class="quota-cell cell--medium cell--mobile"><span class="text is-small is-txtSoftGrey">
        932 
      </span></div>
    <div class="quota-cell cell--medium"><span class="text is-small is-txtSoftGrey">
        1,000 
      </span></div>
    <div class="quota-help cell--small"><a href="https://test-help-link.com" target="_blank"><img svg-inline="" src="../../../assets/icons/common/question-mark.svg" class="quota-image"></a></div>
  </div>
</div>
`;

exports[`QuotaWidget.vue should render warning bar if used ammount is over 60% and under 90% 1`] = `
<div class="quota-widget" quotatype="test">
  <div class="quota-main">
    <div class="quota-cell cell--title">
      <h4 class="text is-caption is-semibold is-txtGrey">My test</h4>
    </div>
    <div class="quota-cell cell--large">
      <div class="progressbar">
        <div class="progressbar progressbar--warning" style="width: 61%;"></div>
      </div>
    </div>
  </div>
  <div class="quota-data">
    <div class="quota-cell cell--medium"><span class="text is-small is-txtSoftGrey">
        390 
      </span></div>
    <div class="quota-cell cell--medium cell--mobile"><span class="text is-small is-txtSoftGrey">
        610 
      </span></div>
    <div class="quota-cell cell--medium"><span class="text is-small is-txtSoftGrey">
        1,000 
      </span></div>
    <div class="quota-help cell--small"><a href="https://test-help-link.com" target="_blank"><img svg-inline="" src="../../../assets/icons/common/question-mark.svg" class="quota-image"></a></div>
  </div>
</div>
`;
